File: apps/desktop/src/platform/host/ai.ts
import type {
  AiModelInfo,
  AiProfileTestResult,
  AiSettingsSnapshot,
  AiTranslateBatchRequest,
  AiTranslateBatchResult,
  AiTranslationCacheEntry,
  AiTranslationCacheStats,
  AiTranslationProgressPayload,
  AiTranslationStreamPayload,
  SaveAiSettingsRequest,
  ExportAiProfilesRequest,
  AiProfileImportConflictPolicy,
  AiProfileImportPreview,
  AiProfileImportResult,
  ModelsDevCatalog,
} from '@shared/contracts'
import { HOST_COMMANDS } from '@platform/host-commands'
import { getPlatformPorts, invokeDesktop } from './runtime'

const AI_PROGRESS_EVENT = 'ai://translation-progress'
const AI_STREAM_EVENT = 'ai://translation-stream'

/** Loads sanitized AI profiles and provider presets without exposing credentials. */
export function loadAiSettings() {
  return invokeDesktop<AiSettingsSnapshot>(HOST_COMMANDS.loadAiSettings, undefined, {
    kind: 'parallelPool',
    pool: 'settings-read',
    limit: 4,
  })
}

/** Persists AI profiles and credential patches serially. */
export function saveAiSettings(request: SaveAiSettingsRequest) {
  return invokeDesktop<AiSettingsSnapshot>(HOST_COMMANDS.saveAiSettings, { request }, { kind: 'queuedMutation', queue: 'AiSettings' })
}

/** Exports sanitized provider profiles; credentials never enter the document. */
export function exportAiProfiles(request: ExportAiProfilesRequest) {
  return invokeDesktop<number>(HOST_COMMANDS.exportAiProfiles, { request }, { kind: 'exclusiveMutation', resource: 'AiProfileExport' })
}

/** Parses and validates a profile document without mutating settings. */
export function previewAiProfilesImport(sourcePath: string) {
  return invokeDesktop<AiProfileImportPreview>(
    HOST_COMMANDS.previewAiProfilesImport,
    { request: { sourcePath } },
    { kind: 'latest', key: `ai-profile-import:${sourcePath}` },
  )
}

/** Applies an already previewed profile document through Rust settings validation. */
export function applyAiProfilesImport(sourcePath: string, conflictPolicy: AiProfileImportConflictPolicy) {
  return invokeDesktop<AiProfileImportResult>(
    HOST_COMMANDS.applyAiProfilesImport,
    { request: { sourcePath, conflictPolicy } },
    { kind: 'exclusiveMutation', resource: 'AiSettings' },
  )
}

/** Lists models exposed by a configured provider profile. */
export function listAiModels(profileId: string) {
  return invokeDesktop<AiModelInfo[]>(
    HOST_COMMANDS.listAiModels,
    { request: { profileId } },
    { kind: 'keyedLatest', key: `ai-models:${profileId}` },
  )
}

/**
 * Fetches the models.dev catalog through the backend Network lane with a
 * memory/disk TTL cache, so repeated dialog opens never re-download the full
 * catalog. Keyed latest keeps concurrent opens on a single in-flight request.
 */
export function fetchAiModelsDevCatalog() {
  return invokeDesktop<ModelsDevCatalog>(HOST_COMMANDS.fetchAiModelsDevCatalog, undefined, {
    kind: 'keyedLatest',
    key: 'ai-models-dev-catalog',
  })
}

/** Executes a small end-to-end inference probe for one saved profile. */
export function testAiProfile(profileId: string) {
  return invokeDesktop<AiProfileTestResult>(
    HOST_COMMANDS.testAiProfile,
    { request: { profileId } },
    { kind: 'serviceGate', key: `ai-test:${profileId}` },
  )
}

/** Runs a bounded translation batch through the dedicated backend AI pool. */
export function translateAiBatch(request: AiTranslateBatchRequest) {
  return invokeDesktop<AiTranslateBatchResult>(
    HOST_COMMANDS.translateAiBatch,
    { request },
    { kind: 'parallelPool', pool: 'ai-translation', limit: 2 },
  )
}

/** Requests cooperative cancellation without queueing behind the active translation. */
export function cancelAiJob(jobId: string) {
  return invokeDesktop<void>(HOST_COMMANDS.cancelAiJob, { request: { jobId } }, { kind: 'serviceGate', key: `ai-cancel:${jobId}` })
}

/** Subscribes to backend translation progress events. */
export function listenToAiProgress(listener: (payload: AiTranslationProgressPayload) => void) {
  return getPlatformPorts().hostEvents.listen<AiTranslationProgressPayload>(AI_PROGRESS_EVENT, listener)
}

/** Subscribes to backend streaming translation deltas (content + reasoning). */
export function listenToAiStream(listener: (payload: AiTranslationStreamPayload) => void) {
  return getPlatformPorts().hostEvents.listen<AiTranslationStreamPayload>(AI_STREAM_EVENT, listener)
}

export function readAiTranslationCache(request: Pick<AiTranslationCacheEntry, 'scopeKey' | 'targetLocale' | 'sourceHash'>) {
  return invokeDesktop<AiTranslationCacheEntry | null>(
    HOST_COMMANDS.readAiTranslationCache,
    { request },
    { kind: 'keyedLatest', key: `ai-cache:${request.scopeKey}:${request.targetLocale}` },
  )
}

export function writeAiTranslationCache(entry: AiTranslationCacheEntry) {
  return invokeDesktop<AiTranslationCacheEntry>(
    HOST_COMMANDS.writeAiTranslationCache,
    { entry },
    { kind: 'queuedMutation', queue: 'AiTranslationCache' },
  )
}

export function getAiTranslationCacheStats() {
  return invokeDesktop<AiTranslationCacheStats>(HOST_COMMANDS.getAiTranslationCacheStats, undefined, {
    kind: 'latest',
    key: 'ai-cache-stats',
  })
}

export function clearAiTranslationCache() {
  return invokeDesktop<AiTranslationCacheStats>(HOST_COMMANDS.clearAiTranslationCache, undefined, {
    kind: 'exclusiveMutation',
    resource: 'AiTranslationCache',
  })
}

File: apps/desktop/src-tauri/src/commands/ai.rs
use crate::domain::ai::types::{
    AiModelInfo, AiProfileImportPreview, AiProfileImportResult, AiProfileRequest,
    AiProfileTestResult, AiSettingsSnapshot, AiTranslateBatchRequest, AiTranslateBatchResult,
    AiTranslationCacheEntry, AiTranslationCacheStats, ApplyAiProfilesImportRequest,
    CancelAiJobRequest, ExportAiProfilesRequest, ModelsDevCatalog, PreviewAiProfilesImportRequest,
    ReadAiTranslationCacheRequest, SaveAiSettingsRequest,
};
use crate::support::logging::DebugLoggingState;
use crate::{AppHandle, AppRuntime};
use serde_json::json;
use tauri::State;

macro_rules! execute {
    ($app:expr, $debug:expr, $name:ident, $args:expr) => {
        crate::commands::runtime::execute_tauri_command(
            AppHandle::from_tauri($app),
            $debug.inner().clone(),
            crate::host_command_name!($name),
            $args,
        )
        .await
    };
}

#[tauri::command]
pub async fn load_ai_settings(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
) -> Result<AiSettingsSnapshot, String> {
    execute!(app, debug, load_ai_settings, json!({}))
}

#[tauri::command]
pub async fn save_ai_settings(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: SaveAiSettingsRequest,
) -> Result<AiSettingsSnapshot, String> {
    execute!(app, debug, save_ai_settings, json!({ "request": request }))
}

#[tauri::command]
pub async fn export_ai_profiles(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: ExportAiProfilesRequest,
) -> Result<u32, String> {
    execute!(
        app,
        debug,
        export_ai_profiles,
        json!({ "request": request })
    )
}

#[tauri::command]
pub async fn preview_ai_profiles_import(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: PreviewAiProfilesImportRequest,
) -> Result<AiProfileImportPreview, String> {
    execute!(
        app,
        debug,
        preview_ai_profiles_import,
        json!({ "request": request })
    )
}

#[tauri::command]
pub async fn apply_ai_profiles_import(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: ApplyAiProfilesImportRequest,
) -> Result<AiProfileImportResult, String> {
    execute!(
        app,
        debug,
        apply_ai_profiles_import,
        json!({ "request": request })
    )
}

#[tauri::command]
pub async fn list_ai_models(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: AiProfileRequest,
) -> Result<Vec<AiModelInfo>, String> {
    execute!(app, debug, list_ai_models, json!({ "request": request }))
}

#[tauri::command]
pub async fn fetch_ai_models_dev_catalog(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
) -> Result<ModelsDevCatalog, String> {
    execute!(app, debug, fetch_ai_models_dev_catalog, json!({}))
}

#[tauri::command]
pub async fn test_ai_profile(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: AiProfileRequest,
) -> Result<AiProfileTestResult, String> {
    execute!(app, debug, test_ai_profile, json!({ "request": request }))
}

#[tauri::command]
pub async fn translate_ai_batch(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: AiTranslateBatchRequest,
) -> Result<AiTranslateBatchResult, String> {
    execute!(
        app,
        debug,
        translate_ai_batch,
        json!({ "request": request })
    )
}

#[tauri::command]
pub async fn cancel_ai_job(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: CancelAiJobRequest,
) -> Result<(), String> {
    execute!(app, debug, cancel_ai_job, json!({ "request": request }))
}

#[tauri::command]
pub async fn read_ai_translation_cache(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    request: ReadAiTranslationCacheRequest,
) -> Result<Option<AiTranslationCacheEntry>, String> {
    execute!(
        app,
        debug,
        read_ai_translation_cache,
        json!({ "request": request })
    )
}

#[tauri::command]
pub async fn write_ai_translation_cache(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
    entry: AiTranslationCacheEntry,
) -> Result<AiTranslationCacheEntry, String> {
    execute!(
        app,
        debug,
        write_ai_translation_cache,
        json!({ "entry": entry })
    )
}

#[tauri::command]
pub async fn get_ai_translation_cache_stats(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
) -> Result<AiTranslationCacheStats, String> {
    execute!(app, debug, get_ai_translation_cache_stats, json!({}))
}

#[tauri::command]
pub async fn clear_ai_translation_cache(
    app: tauri::AppHandle<AppRuntime>,
    debug: State<'_, DebugLoggingState>,
) -> Result<AiTranslationCacheStats, String> {
    execute!(app, debug, clear_ai_translation_cache, json!({}))
}
